### PR TITLE
refactor: error handling, login cleanup, and test coverage

### LINKED ISSUE #246: refactor: long function + SRP — break LoginHandler into focused helpers
## Summary

The `LoginHandler` function (tools/login.go:74-304) is 231 lines long — a "god function" handling navigation, trigger detection, two parallel credential-filling flows, form submission, success verification, and result assembly. This makes it hard to test, debug, and extend.

## Location

`tools/login.go:74-304`

## Category & Severity

- **Category**: 3 (Long Functions) + 12 (SRP Violations)
- **Severity**: P2 (High)

## Evidence

```go
LoginHandler = func(rodCtx *types.Context) server.ToolHandlerFunc {
    handler := func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
        // Line 76-107: Parse 10+ arguments
        // Line 109-120: Navigate and wait
        // Line 122-187: Trigger-selector path (modal login)
        // Line 188-233: Standard path (page login)
        // Line 235-272: Success verification polling loop
        // Line 274-301: Result assembly
    }
}
```

## Suggested Refactoring

Break into focused helper functions:

1. `fillLoginForm(page, selectors, credentials, formContainer) error` — consolidates both trigger and standard fill paths
2. `verifyLoginSuccess(page, timeout, successSelector, successURL) (bool, error)` — extracts polling loop
3. `buildLoginResult(page, verified, timeout) (map[string]any, error)` — result assembly

The trigger vs standard split at line 122 shares ~80% identical logic (find username → fill → find password → fill → find submit → click). Extract the shared sequence, parameterize the scope.

## Effort Estimate

- **Size**: Medium (~2-3 hours)
- **Risk**: Low (internal refactor, no API change)
- **Tests**: Existing E2E tests cover the login flow

## Acceptance Criteria

- [ ] `LoginHandler` body is <50 lines, delegating to named helpers
- [ ] Trigger and standard paths share the form-filling logic
- [ ] All existing E2E login tests pass unchanged
- [ ] No new exports added (helpers are unexported)

### LINKED ISSUE #247: refactor: DRY — deduplicate login credential-filling between trigger and standard paths
## Summary

The login credential-filling sequence is duplicated between the trigger-selector path (lines 138-187) and standard path (lines 189-233). Both paths perform identical operations: find username → smart fill → find password → smart fill → find submit → click. The only difference is CSS selector scoping.

## Location

- `tools/login.go:138-162` (trigger path username fill)
- `tools/login.go:189-214` (standard path username fill)
- `tools/login.go:164-187` (trigger path password+submit)
- `tools/login.go:216-233` (standard path password+submit)

## Category & Severity

- **Category**: 2 (DRY Violations)
- **Severity**: P2 (High)

## Evidence

```go
// Trigger path (lines 138-162):
if userSelector == "" {
    filled := false
    for _, sel := range defaultUsernameSelectors {
        scopedSel := formContainer + " " + sel
        el, elErr := page.Element(scopedSel)
        if elErr == nil {
            if fillErr := loginSmartFill(el, username); fillErr == nil {
                filled = true; break
            }
        }
    }
}

// Standard path (lines 200-213) — near-identical:
filled := false
for _, sel := range defaultUsernameSelectors {
    el, err := page.Element(sel)  // only difference: no scope prefix
    if err == nil {
        if fillErr := loginSmartFill(el, username); fillErr == nil {
            filled = true; break
        }
    }
}
```

~95 lines duplicated with the only difference being optional `formContainer` prefix on selectors.

## Suggested Refactoring

```go
type loginFormOpts struct {
    page           *rod.Page
    scope          string // e.g. "[role=dialog]" or ""
    userSelector   string
    passSelector   string
    submitSelector string
    username       string
    password       string
}

func fillAndSubmitLoginForm(opts loginFormOpts) error {
    scopePrefix := ""
    if opts.scope != "" {
        scopePrefix = opts.scope + " "
    }
    // unified find-fill-submit logic
}
```

## Effort Estimate

- **Size**: Small (~1 hour)
- **Risk**: Low
- **Tests**: E2E tests cover both paths

## Acceptance Criteria

- [ ] Single `fillAndSubmitLoginForm` function handles both trigger and standard paths
- [ ] Selector scoping handled via optional scope parameter
- [ ] No duplicated element-finding or filling logic
- [ ] E2E tests pass unchanged

### LINKED ISSUE #249: refactor: error context — wrap bare error returns in initLocked and closeBrowser
## Summary

Multiple bare `return err` statements in `initLocked()` lose call-chain context. When browser launch or page creation fails, the error propagates without indicating which initialization step failed or what was being attempted.

## Location

- `types/context.go:131-138` (`initLocked` — 4 bare returns)
- `types/context.go:288` (`closeBrowser` — 1 bare return)
- `types/context_tabs.go:92` (1 bare return)

## Category & Severity

- **Category**: 15 (Error Context Propagation)
- **Severity**: P2 (High)

## Evidence

```go
// types/context.go:127-149
func (ctx *Context) initLocked() error {
    var err error
    if ctx.browser == nil {
        ctx.browser, ctx.clonedProfileDir, err = launchBrowser(ctx.stdContext, ctx.config)
        if err != nil {
            return err  // bare — lost: "what was being initialized?"
        }
        ctx.startKeepalive()
        ctx.page, err = ctx.createPage()
        if err != nil {
            return err  // bare — lost: "page creation after browser launch"
        }
        return nil
    }
    if ctx.page == nil {
        ctx.page, err = ctx.createPage()
        if err != nil {
            return err  // bare — lost: "page creation for existing browser"
        }
    }
    return nil
}
```

When `launchBrowser` returns an error, the caller sees the raw error with no indication that it came from browser initialization. This makes debugging startup failures harder.

## Suggested Refactoring

```go
func (ctx *Context) initLocked() error {
    var err error
    if ctx.browser == nil {
        ctx.browser, ctx.clonedProfileDir, err = launchBrowser(ctx.stdContext, ctx.config)
        if err != nil {
            return fmt.Errorf("init browser: %w", err)
        }
        ctx.startKeepalive()
        ctx.page, err = ctx.createPage()
        if err != nil {
            return fmt.Errorf("init page after browser launch: %w", err)
        }
        return nil
    }
    if ctx.page == nil {
        ctx.page, err = ctx.createPage()
        if err != nil {
            return fmt.Errorf("init page for existing browser: %w", err)
        }
    }
    return nil
}
```

Similarly for `closeBrowser`:
```go
err := ctx.closePage()
if err != nil {
    return fmt.Errorf("close browser: close page: %w", err)
}
```

## Effort Estimate

- **Size**: Small (~30 min)
- **Risk**: None (additive, no behavior change)
- **Tests**: Verify error messages in existing tests

## Acceptance Criteria

- [ ] All `return err` in `initLocked`, `closeBrowser`, and tab functions wrap with `fmt.Errorf("context: %w", err)`
- [ ] Error messages indicate which step failed
- [ ] No behavior changes — only error strings enriched

### LINKED ISSUE #250: refactor: missing error handling — stop ignoring marshal and page.Info errors
## Summary

Several tool handlers ignore `json.MarshalIndent` errors with `_ =`, and `page.Info()` results are used without error checking. While marshal errors on `map[string]interface{}` are rare, ignoring them means silent data loss. The `page.Info()` pattern is more concerning since it can fail if the page was navigated away.

## Location

- `tools/login.go:275` — `info, _ := page.Info()`
- `tools/login.go:284` — `resp, _ := proto.NetworkGetCookies{}.Call(page)`
- `tools/login.go:300` — `out, _ := json.MarshalIndent(result, "", "  ")`
- Similar pattern in other tool handlers (search for `_, _ :=` and `out, _ :=`)

## Category & Severity

- **Category**: 7 (Missing Error Handling)
- **Severity**: P2 (High)

## Evidence

```go
// tools/login.go:275-300
info, _ := page.Info()           // can fail if page crashed/navigated
currentURL := ""
title := ""
if info != nil {
    currentURL = info.URL
    title = info.Title
}

resp, _ := proto.NetworkGetCookies{}.Call(page)  // can fail if session dead
cookieCount := 0
if resp != nil {
    cookieCount = len(resp.Cookies)
}

// ...
out, _ := json.MarshalIndent(result, "", "  ")  // can fail on unusual types
return mcp.NewToolResultText(string(out)), nil   // nil bytes → empty string
```

## Suggested Refactoring

```go
info, infoErr := page.Info()
if infoErr != nil {
    log.Debugf("login result: page info unavailable: %v", infoErr)
}
// ... handle nil info as before

out, err := json.MarshalIndent(result, "", "  ")
if err != nil {
    return toolErr("login marshal result", err)
}
```

## Effort Estimate

- **Size**: Small (~1 hour)
- **Risk**: None
- **Tests**: Verify error paths in unit tests

## Acceptance Criteria

- [ ] No `_ =` on error returns in production code (grep for `_, _ :=` and `_ =.*err`)
- [ ] `json.MarshalIndent` errors returned or logged
- [ ] `page.Info()` errors at least logged at debug level
- [ ] `proto.NetworkGetCookies` failure handled gracefully

### LINKED ISSUE #251: refactor: error context — add debug logging to walkIframeNode silent failures
## Summary

Errors in `walkIframeNode` are silently swallowed — when iframe element query, frame access, or nested snapshot capture fail, the function returns a fallback `<could not capture iframe snapshot>` node with zero logging. This makes iframe snapshot failures invisible during debugging.

## Location

`types/snapshot.go:196-224`

## Category & Severity

- **Category**: 15 (Error Context Propagation)
- **Severity**: P2 (High)

## Evidence

```go
func (s *Snapshot) walkIframeNode(node *yaml.Node, frame *rod.Page) *yaml.Node {
    matches := refPattern.FindStringSubmatch(node.Value)
    if len(matches) <= 1 {
        return nil
    }

    ref := matches[1]
    fallback := &yaml.Node{Kind: yaml.ScalarNode, Value: "<could not capture iframe snapshot>"}
    pairNode := &yaml.Node{
        Kind:    yaml.MappingNode,
        Content: []*yaml.Node{{Kind: yaml.ScalarNode, Value: node.Value}, fallback},
    }

    childFrameEle, err := utils.QueryEleByAria(frame, ref)
    if err != nil {
        return pairNode  // ← silent: why did query fail?
    }
    childFrame, err := childFrameEle.Frame()
    if err != nil {
        return pairNode  // ← silent: why couldn't get frame?
    }
    childSnapshot, err := s.captureSnapshotWithFrames(childFrame)
    if err != nil || len(childSnapshot.Content) == 0 {
        return pairNode  // ← silent: why did snapshot fail?
    }

    pairNode.Content[1] = childSnapshot.Content[0]
    return pairNode
}
```

Three error paths all return the same fallback node with no indication of what went wrong. When users see `<could not capture iframe snapshot>` in their output, there's no way to diagnose the cause.

## Suggested Refactoring

Add debug-level logging at each failure point:

```go
childFrameEle, err := utils.QueryEleByAria(frame, ref)
if err != nil {
    log.Debugf("iframe snapshot: query element by aria ref=%q: %v", ref, err)
    return pairNode
}
childFrame, err := childFrameEle.Frame()
if err != nil {
    log.Debugf("iframe snapshot: get frame for ref=%q: %v", ref, err)
    return pairNode
}
childSnapshot, err := s.captureSnapshotWithFrames(childFrame)
if err != nil {
    log.Debugf("iframe snapshot: capture nested snapshot for ref=%q: %v", ref, err)
    return pairNode
}
```

## Effort Estimate

- **Size**: Small (~20 min)
- **Risk**: None (additive logging only)
- **Tests**: None needed

## Acceptance Criteria

- [ ] Each error path in `walkIframeNode` logs at debug level with the ref and error
- [ ] Fallback behavior unchanged — still returns `<could not capture iframe snapshot>`
- [ ] No new imports beyond existing `log` package

### LINKED ISSUE #252: refactor: test coverage — add unit tests for login form detection and filling
## Summary

The login tool has no unit tests for its core form-detection and credential-filling logic. The `loginSmartFill` function, default selector fallback, trigger-vs-standard path selection, and success verification polling loop are only covered by E2E tests. Unit tests would catch regressions faster and make the upcoming refactoring safer.

## Location

- `tools/login.go:44-51` (default selectors — untested fallback ordering)
- `tools/login.go:54-71` (`loginSmartFill` — React-aware fill, untested)
- `tools/login.go:235-272` (success verification loop — untested timeout behavior)

## Category & Severity

- **Category**: 18 (Test Coverage Gaps)
- **Severity**: P2 (High)

## Evidence

No `tools/login_test.go` file exists. The login handler contains:
- Complex branching (trigger vs standard × user-supplied vs default selectors = 4 paths)
- Timeout-based polling with `time.After` and `ticker`
- React state synchronization via JS eval
- Error handling with graceful degradation

All this is currently only tested via E2E browser tests which are slow and don't exercise edge cases (timeout races, partial fill failures, malformed smart-fill responses).

## Suggested Refactoring

After extracting helpers (see related LoginHandler refactor issue), add unit tests:

```go
// tools/login_test.go
func TestLoginSmartFill_ParsesSuccessResult(t *testing.T) { ... }
func TestLoginSmartFill_HandlesParseFailure(t *testing.T) { ... }
func TestLoginSmartFill_ReportsValueMismatch(t *testing.T) { ... }
func TestDefaultUsernameSelectors_AreValid(t *testing.T) { ... }
func TestVerifyLoginSuccess_TimesOut(t *testing.T) { ... }
func TestVerifyLoginSuccess_MatchesURLSubstring(t *testing.T) { ... }
```

## Effort Estimate

- **Size**: Medium (~2-3 hours, depends on helper extraction)
- **Risk**: None (additive)
- **Tests**: This IS the test work

## Acceptance Criteria

- [ ] `tools/login_test.go` exists with ≥6 test functions
- [ ] Smart fill success/failure/parse-error paths tested
- [ ] Success verification timeout behavior tested
- [ ] Default selector list validated (non-empty, valid CSS)

### LINKED ISSUE #253: refactor: SRP — extract walkScalarNode into focused helper methods
## Summary

`walkScalarNode` performs three distinct operations in a single pass: frame ref adjustment, iframe snapshot expansion, and ref index accumulation. While combining them into one pass is intentional for performance, the function's responsibilities are tangled and hard to test independently.

## Location

`types/snapshot.go:167-192`

## Category & Severity

- **Category**: 12 (SRP Violations)
- **Severity**: P3 (Medium)

## Evidence

```go
func (s *Snapshot) walkScalarNode(node *yaml.Node, frameIndex int, frame *rod.Page) (*yaml.Node, error) {
    if node.Tag != "!!str" {
        return node, nil
    }

    value := node.Value
    // Operation 1: Frame ref adjustment
    if frameIndex > 0 {
        node.Value = strings.Replace(value, "[ref=", fmt.Sprintf("[ref=f%d", frameIndex), 1)
    }

    // Operation 2: Iframe expansion
    if strings.HasPrefix(value, "iframe ") {
        if result := s.walkIframeNode(node, frame); result != nil {
            return result, nil
        }
    }

    // Operation 3: Ref accumulation
    if s.refAccum != nil {
        effectiveValue := node.Value
        if entry, ok := parseRefScalar(effectiveValue); ok {
            *s.refAccum = append(*s.refAccum, entry)
        }
    }

    return node, nil
}
```

## Suggested Refactoring

Extract named methods for clarity while keeping the single-pass optimization:

```go
func (s *Snapshot) walkScalarNode(node *yaml.Node, frameIndex int, frame *rod.Page) (*yaml.Node, error) {
    if node.Tag != "!!str" {
        return node, nil
    }
    s.adjustFrameRef(node, frameIndex)
    if result := s.tryExpandIframe(node, frame); result != nil {
        return result, nil
    }
    s.accumulateRef(node.Value)
    return node, nil
}

func (s *Snapshot) adjustFrameRef(node *yaml.Node, frameIndex int) { ... }
func (s *Snapshot) tryExpandIframe(node *yaml.Node, frame *rod.Page) *yaml.Node { ... }
func (s *Snapshot) accumulateRef(value string) { ... }
```

## Effort Estimate

- **Size**: Small (~30 min)
- **Risk**: Low
- **Tests**: Existing snapshot tests

## Acceptance Criteria

- [ ] `walkScalarNode` body is ≤10 lines, delegating to named helpers
- [ ] Each helper is independently testable
- [ ] Snapshot output unchanged (verify via existing tests)

### LINKED ISSUE #254: refactor: config externalization — make login selector defaults configurable
## Summary

Default login selectors (`defaultUsernameSelectors`, default password/submit/form-container selectors) are hardcoded in `tools/login.go`. Sites with non-standard forms require code changes. These should be configurable via the YAML config file.

## Location

- `tools/login.go:44-51` (defaultUsernameSelectors)
- `tools/login.go:99-107` (hardcoded defaults for password, submit, form container)

## Category & Severity

- **Category**: 11 (Config Externalization)
- **Severity**: P3 (Medium)

## Evidence

```go
// tools/login.go:44-51 — hardcoded, not configurable
var defaultUsernameSelectors = []string{
    "input[type=email]",
    "input[name=email]",
    "input[name=username]",
    "input[name=login]",
    "input[id=email]",
    "input[id=username]",
}

// tools/login.go:99-107 — more hardcoded defaults
if passSelector == "" {
    passSelector = "input[type=password]"
}
if submitSelector == "" {
    submitSelector = "button[type=submit]"
}
if formContainer == "" {
    formContainer = "[role=dialog]"
}
```

## Suggested Refactoring

Add to `types/config.go`:

```go
type LoginDefaults struct {
    UsernameSelectors []string `yaml:"usernameSelectors,omitempty"`
    PasswordSelector  string   `yaml:"passwordSelector,omitempty"`
    SubmitSelector    string   `yaml:"submitSelector,omitempty"`
    FormContainer     string   `yaml:"formContainer,omitempty"`
}
```

In the login handler, fall back to code defaults when config is empty:

```go
selectors := rodCtx.Config().LoginDefaults.UsernameSelectors
if len(selectors) == 0 {
    selectors = defaultUsernameSelectors
}
```

## Effort Estimate

- **Size**: Small (~1 hour)
- **Risk**: Low (additive config, existing defaults preserved)
- **Tests**: Add config parsing test for new fields

## Acceptance Criteria

- [ ] Login selector defaults configurable via YAML config
- [ ] Empty config falls back to current hardcoded defaults
- [ ] Config documented in rod-mcp.yaml example
- [ ] Existing behavior unchanged when config not provided

### LINKED ISSUE #255: refactor: magic numbers — consolidate scattered timeout constants
## Summary

While `tools/constants.go` consolidates most timeout defaults, some timeouts remain scattered in other files as local constants or magic numbers. The login timeout, keepalive interval, keepalive ping timeout, and a 5-second form-container wait are defined separately.

## Location

- `tools/login.go:23` — `defaultLoginTimeoutMs = 15000.0`
- `tools/login.go:132` — `page.Timeout(5 * time.Second)` (magic number)
- `types/context.go:306` — `keepaliveInterval = 5 * time.Minute`
- `types/context.go:334` — `context.WithTimeout(stopCtx, 10*time.Second)` (magic number)
- `tools/assert.go:60` — `time.NewTicker(200 * time.Millisecond)` (magic polling interval)

## Category & Severity

- **Category**: 6 (Magic Numbers/Strings)
- **Severity**: P3 (Medium)

## Evidence

```go
// tools/login.go:23
const defaultLoginTimeoutMs = 15000.0  // not in constants.go

// tools/login.go:132 — inline magic number
if _, err := page.Timeout(5 * time.Second).Element(formContainer); err != nil {

// types/context.go:306 — separate const block
const keepaliveInterval = 5 * time.Minute

// types/context.go:334 — inline magic number
pingCtx, pingCancel := context.WithTimeout(stopCtx, 10*time.Second)

// tools/assert.go:60 — inline magic number
ticker := time.NewTicker(200 * time.Millisecond)
```

## Suggested Refactoring

1. Move `defaultLoginTimeoutMs` to `tools/constants.go`
2. Extract magic numbers as named constants:

```go
// tools/constants.go additions:
const (
    defaultLoginTimeoutMs        = 15000.0
    defaultFormContainerTimeout  = 5 * time.Second
    defaultPollInterval          = 200 * time.Millisecond
)

// types/constants.go (new, or add to existing):
const (
    keepaliveInterval    = 5 * time.Minute
    keepalivePingTimeout = 10 * time.Second
)
```

3. Replace inline numbers with named constants

## Effort Estimate

- **Size**: Small (~30 min)
- **Risk**: None
- **Tests**: None needed (value-preserving rename)

## Acceptance Criteria

- [ ] All timeout/interval values defined as named constants
- [ ] No magic time.Duration or float64 timeout values in function bodies
- [ ] Related constants grouped together in constant blocks

### PR DESCRIPTION
**Pull request by aliwatters**
## Summary

Slot 10 refactoring batch — 9 issues addressing error handling, login code quality, and test coverage.

## Issues

| Issue | Change |
|-------|--------|
| #255 | Consolidate login timeout constant into constants.go |
| #249 | Wrap bare error returns in initLocked and closeBrowser |
| #250 | Handle ignored marshal and page.Info errors across 8 files |
| #251 | Add debug logging to walkIframeNode silent failures |
| #253 | Extract walkScalarNode into focused helper methods |
| #246 | Break LoginHandler into focused helpers (~40 lines of orchestration) |
| #247 | Deduplicate login credential-filling between trigger and standard paths |
| #254 | Make login selector defaults configurable via Config |
| #252 | Add unit tests for login parameter parsing and form detection |

## Key changes

- **LoginHandler**: Refactored from a 230-line monolith to ~40 lines of orchestration calling 7 focused helpers (`parseLoginParams`, `loginFillUsername`, `loginFillPassword`, `loginSubmit`, `loginOpenTrigger`, `loginVerifySuccess`, `loginBuildResult`)
- **Error handling**: All `json.MarshalIndent` and `page.Info()` ignored errors now handled (logged or returned)
- **Error context**: `initLocked` and `closeBrowser` wrap errors with `fmt.Errorf` for traceability
- **Observability**: `walkIframeNode` failures logged at debug level
- **Config**: Login selectors configurable via `loginUsernameSelectors`, `loginPasswordSelector`, `loginSubmitSelector`
- **Tests**: 10 new unit tests for login parameter parsing, config overrides, and defaults

## Test Plan

- [x] `go vet ./...` passes
- [x] `go build ./...` passes
- [x] `go test ./...` passes (including 10 new login tests)
- [x] E2E: `TestE2E_Navigation` — all subtests pass including 404 fast-fail
- [x] E2E: `TestE2E_Login` — 404 fast-fail passes

Fixes #246, fixes #247, fixes #249, fixes #250, fixes #251, fixes #252, fixes #253, fixes #254, fixes #255

### CODE REVIEW

File: tools/login.go [hunk 1]
```diff
@@ -70,48 +68,226 @@ func loginSmartFill(element *rod.Element, value string) error {
...
+	triggerEl, err := page.Element(triggerSelector)
+	if err != nil {
+		return fmt.Errorf("find trigger %q: %w", triggerSelector, err)
+	}
+	if err := triggerEl.Click(proto.InputMouseButtonLeft, 1); err != nil {
+		return fmt.Errorf("click trigger: %w", err)
+	}
+	if _, err := page.Timeout(5 * time.Second).Element(formContainer); err != nil {
+		return fmt.Errorf("form container %q did not appear after clicking trigger: %w", formContainer, err)
+	}
```
[the remark this answers is not in the document]
Comment (aliwatters) [reply tools/login.go]: Fixed in 2461c4b — extracted to defaultFormContainerTimeout constant in constants.go.

File: tools/login.go [hunk 2]
```diff
@@ -70,48 +68,226 @@ func loginSmartFill(element *rod.Element, value string) error {
...
+}
+
+// loginBuildResult collects the login result (URL, title, cookies) into JSON.
+func loginBuildResult(page *rod.Page, verified bool, timeout float64) (string, error) {
+	info, infoErr := page.Info()
+	currentURL := ""
+	title := ""
+	if infoErr != nil {
+		log.Warnf("login page info: %s", infoErr)
```
[the remark this answers is not in the document]
Comment (aliwatters) [reply tools/login.go]: Fixed in 2461c4b — downgraded to debug level to match how other tools treat best-effort page.Info() calls.

File: tools/login.go [hunk 3]
```diff
@@ -70,48 +68,226 @@ func loginSmartFill(element *rod.Element, value string) error {
...
+		log.Warnf("login get cookies: %s", cookieErr)
+	} else if resp != nil {
+		cookieCount = len(resp.Cookies)
+	}
+
+	result := map[string]interface{}{
+		"success":     verified,
+		"url":         currentURL,
+		"title":       title,
+		"cookies_set": cookieCount,
```
[the remark this answers is not in the document]
Comment (aliwatters) [reply tools/login.go:263]: Downgraded to debug level in 2461c4b. Skipping the cookies_retrieved field — adding a new result field for an edge case that only occurs during page teardown would be over-engineering. Debug log is sufficient for diagnosis.
